### PR TITLE
Use a simpler mechanism to detect file changes

### PR DESCRIPTION
**Pull request by keynmol**
This greatly simplifies the caching logic and solves the problem I was having - if you run `~compile` on a module with smithy specs, they actually weren't re-generated and re-compiled if you changed a spec.

I've confirmed on my project that this change addresses the problem.

### CODE REVIEW

File: modules/codegen-plugin/src/sbt-test/codegen-plugin/custom-settings/project/build.properties
Comment (keynmol) [inline modules/codegen-plugin/src/sbt-test/codegen-plugin/custom-settings/project/build.properties:1]: I had to do this to run any scripted tests, otherwise it said this:

```
[error] [error] [launcher] error during sbt launcher: java.lang.UnsatisfiedLinkError: Can't load library: /Users/velvetbaldmime/Library/Caches/JNA/temp/jna4456582555411290288.tmp
[error] x codegen-plugin/defaults
sbt.internal.scripted.TestException: {line -1}  Reload for scripted batch execution failed.
  | => cat sbt.scriptedtest.ScriptedTests.$anonfun$runBatchedTests$5(ScriptedTests.scala:237)
	at sbt.scriptedtest.ScriptedTests.$anonfun$runBatchedTests$5$adapted(ScriptedTests.scala:231)
	at sbt.scriptedtest.ScriptedTests.$anonfun$commonRunTest$6(ScriptedTests.scala:325)
	at scala.util.control.Exception$Catch.apply(Exception.scala:228)
	at sbt.scriptedtest.ScriptedTests.commonRunTest(ScriptedTests.scala:324)
	at sbt.scriptedtest.ScriptedTests.$anonfun$runBatchedTests$3(ScriptedTests.scala:241)
	at sbt.scriptedtest.ScriptedTests.runOrHandleDisabled(ScriptedTests.scala:275)
	at sbt.scriptedtest.ScriptedTests.$anonfun$runBatchedTests$1(ScriptedTests.scala:245)
```

Which I think is a known JNA issue on M1 macs in older versions of SBT.

Which also makes me think none of y'all are running scripted tests locally 🤠

File: .github/workflows/ci.yml
Comment (keynmol) [inline .github/workflows/ci.yml:68]: I don't think scripted tests were running at all - let me know if I missed it somewhere

File: modules/codegen-plugin/src/smithy4s/codegen/CodegenPlugin.scala
Comment (keynmol) [inline modules/codegen-plugin/src/smithy4s/codegen/CodegenPlugin.scala:149]: The general idea is simple - if there were _any_ changes, we just reconstruct the "live" set of smithy files (which means unmodified + modified - deleted)

If there were no changes, we need to return the list of output files still - otherwise SBT won't pick them up for future compilation.

File: modules/codegen-plugin/src/sbt-test/codegen-plugin/custom-settings/project/build.properties
Comment (Baccata) [reply modules/codegen-plugin/src/sbt-test/codegen-plugin/custom-settings/project/build.properties:1]: Considering the SBT plugin is just a small wrapper over the logic held by the `codegen` module, and that we test that via bootstrapping, I don't really care about the scripted tests that much 😄

Comment (Baccata) [reply modules/codegen-plugin/src/sbt-test/codegen-plugin/custom-settings/project/build.properties:1]: but thanks anyway

File: modules/codegen-plugin/src/smithy4s/codegen/CodegenPlugin.scala
Comment (Baccata) [reply modules/codegen-plugin/src/smithy4s/codegen/CodegenPlugin.scala:149]: Would you mind inlining that as a comment in the code ? That's useful information to have when we read the files

**Review by Baccata — APPROVED** [review]
I think it all makes sense. Would like @daddykotex to give his review before we merge

File: modules/codegen-plugin/src/smithy4s/codegen/CodegenPlugin.scala
Comment (daddykotex) [reply modules/codegen-plugin/src/smithy4s/codegen/CodegenPlugin.scala:149]: that's amazing I did not know sbt had `inputFileChanges` along with `fileInputs` and `fileOutputs`

**Review by daddykotex — APPROVED** [review]
Great PR

It looks like we're piggy backing on sbt's to handle that for us, that's great

also thanks for adding the scripted test to the CI

**Comment by daddykotex** [comment]
@Baccata could I add the comment and push a commit on top of @keynmol 's PR?

is that something maintainers do?

**Comment by Baccata** [comment]
> @Baccata could I add the comment and push a commit on top of @keynmol 's PR?
> is that something maintainers do? 

Sure, go ahead.

**Review by lewisjkl — APPROVED** [review]

**Pull request merged**
